processing/data_collection/gazette/spiders/sp_santo_andre.py

```diff
@@ -0,0 +1,187 @@
+import dateparser
+from datetime import datetime, date
+from urllib.parse import unquote
+from gazette.items import Gazette
+from gazette.spiders.base import BaseGazetteSpider
+from scrapy import FormRequest, Request
+from scrapy.selector.unified import Selector
+from gazette.settings import FILES_STORE
+import re
+from w3lib.html import remove_tags
+from pathlib import Path
+from typing import Optional
+from hashlib import sha1
+
+
+class SpSantoAndreSpider(BaseGazetteSpider):
+    name = "sp_santo_andre"
+    TERRITORY_ID = "3547809"
+    custom_settings = {
+        "ITEM_PIPELINES": {
+            "gazette.pipelines.GazetteDateFilteringPipeline": 50,
+            "gazette.pipelines.ExtractTextPipeline": 200,
+        }
+    }
+
+    allowed_domains = ["santoandre.sp.gov.br"]
+    start_urls = [
+        "http://www.santoandre.sp.gov.br/publicacao/edicao/consultaedicao.aspx"
+    ]
+    start_date = date(2010, 1, 1)
+    TODAY = datetime.now().strftime("%d/%m/%Y")
+    JAVASCRIPT_POSTBACK_REGEX = r"javascript:__doPostBack\('(.*)',''\)"
+    CURRENT_PAGE = 1
+
+    def start_requests(self):
+        for url in self.start_urls:
+            yield Request(url, callback=self.search_by_date)
+
+    def _get_form_params(self, response):
+        view_state = response.css("#__VIEWSTATE::attr(value)").get()
+        event_validation = response.css("#__EVENTVALIDATION::attr(value)").get()
+        return view_state, event_validation
+
+    def search_by_date(self, response):
+        VIEW_STATE, EVENT_VALIDATION = self._get_form_params(response)
+        yield FormRequest.from_response(
+            response,
+            callback=self.parse,
+            formname="aspnetForm",
+            formdata={
+                "__EVENTTARGET": "",
+                "__EVENTARGUMENT": "",
+                "__VIEWSTATE": VIEW_STATE,
+                "__EVENTVALIDATION": EVENT_VALIDATION,
+                "ctl00$ContentPlaceHolder1$dt_inicio": self.start_date.strftime(
+                    "%d/%m/%Y"
+                ),
+                "ctl00$ContentPlaceHolder1$dt_final": self.TODAY,
+            },
+            clickdata={"name": "ctl00$ContentPlaceHolder1$PsaToolBar1$btnSelecionar"},
+            method="POST",
+        )
+
+    def _save_pdf(self, response, path: Path, item):
+        if not isinstance(path, Path):
+            raise TypeError("O parâmetro path precisa ser uma instância de Path.")
+        path.parent.mkdir(parents=True, exist_ok=True)
+        with open(path, "wb") as f:
```

Comment:
**issue**: We can't store the file locally. Our pipeline will store files in a S3-compatible space, so this solution is not acceptable. Probably we will need a custom FilePipeline that is able to upload the content of the PDF to a S3 bucket.

Reply:
Perhaps, can I create another method to save these data on S3 bucket with [boto3](https://github.com/boto/boto3). What do you think? Do you have any suggestions? Or maybe we need to wait for some decisions about the infrastructure of the project? Perhaps this is not the best choice because it can freeze many on going new spiders for the project.

Reply:
I would try at first investigate the use of a custom pipeline based on `scrapy.pipelines.files.FilesPipeline` that we use in the project specific for this spider. It would be better to keep spider's code unaware about how the files extracted are stored.

Reply:
I agree with @rennerocha . I thought of a workaround like this, let me know what you guys think.

```python
# pipelines.py
...
from itemadapter import ItemAdapter
from scrapy.http import Request
from scrapy.pipelines.files import FilesPipeline
from scrapy.settings import Settings
...
class QueridoDiarioFilesPipeline(FilesPipeline):
    DEFAULT_FILES_REQUESTS_FIELD = 'file_requests'

    def __init__(self, *args, settings=None, **kwargs):
        super().__init__(*args, settings=settings, **kwargs)

        if isinstance(settings, dict) or settings is None:
            settings = Settings(settings)

        self.files_requests_field = settings.get(
            'FILES_REQUESTS_FIELD', self.DEFAULT_FILES_REQUESTS_FIELD
        )


    def get_media_requests(self, item, info):
        urls = ItemAdapter(item).get(self.files_urls_field, [])
        yield from (Request(u) for u in urls)
        requests = ItemAdapter(item).get(self.files_requests_field, [])
        yield from requests
```
```python
# items.py
...
class Gazette(scrapy.Item):
	...
	file_requests = scrapy.Field()
	...
```

Reply:
> ```python
> self.files_requests_field
> ```

I have to confess that I didn't follow you. That new field on Item, what is his purpose? The traditional field (`files_urls_field`) receives the URLs that will be used by `RequestWithItem`. Furthermore, in our scenario, we need to made POST request via `FormRequest`, right? Thereby, where our POST requests make a relationship with this new field? I had some research, and it seems that we need to send out POST to the scrapy scheduler to process the pipeline, something like that...

Reply:
Sorry for not clearing that up. With a pipeline like that instead of the default one we would be using a `file_requests` field on the `Gazette` item if we needed to send requests instead of `file_urls`. Building the requests instances inside the spider and sending them away inside a list (`file_requests`) through the item to be requested during pipeline processing. I felt like a new field would be better to not mix the terminology.

Oh, and the existing `QueridoDiarioFilesPipeline` is unrelated to this one. Sorry for giving it the same name and mixing those up. I've changed this one's name to make it clear.